retrieval/discovery/azure.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/arm/compute"
+	"github.com/Azure/azure-sdk-for-go/arm/network"
+	"github.com/Azure/go-autorest/autorest/azure"
+
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/util/strutil"
+)
+
+const (
+	azureLabel                     = model.MetaLabelPrefix + "azure_"
+	azureLabelMachineID            = azureLabel + "machine_id"
+	azureLabelMachineResourceGroup = azureLabel + "machine_resource_group"
+	azureLabelMachineName          = azureLabel + "machine_name"
+	azureLabelMachineLocation      = azureLabel + "machine_location"
+	azureLabelMachinePrivateIP     = azureLabel + "machine_private_ip"
+	azureLabelMachineTag           = azureLabel + "machine_tag_"
+)
+
+// AzureDiscovery periodically performs Azure-SD requests. It implements
+// the TargetProvider interface.
+type AzureDiscovery struct {
+	cfg      *config.AzureSDConfig
+	interval time.Duration
+	port     int
+}
+
+// NewAzureDiscovery returns a new AzureDiscovery which periodically refreshes its targets.
+func NewAzureDiscovery(cfg *config.AzureSDConfig) *AzureDiscovery {
+	return &AzureDiscovery{
+		cfg:      cfg,
+		interval: time.Duration(cfg.RefreshInterval),
+		port:     cfg.Port,
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (ad *AzureDiscovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(ad.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	tg, err := ad.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		ch <- []*config.TargetGroup{tg}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := ad.refresh()
+			if err != nil {
+				log.Error(err)
+			} else {
+				ch <- []*config.TargetGroup{tg}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
```

Comment:
>Does that explanation make sense?

Perfectly, thank you. I didn't realize ```case <-ticker.C``` would block. In this case, it appears to be basically a sleep if I'm understanding this correctly.

I was under the impression that it would only fire when the tick was up and move on otherwise but now that I think about it, it must block.

Thanks for the explanation. Today is being consumed by production issues but I'll try to get to updating this PR as soon as I am able.